### PR TITLE
small fixes in fusion_compiler

### PR DESCRIPTION
**Pull request by ngimel**
Don't call cudaFree unconditionally, guard cudaFree call on cudaFreeMutex, submit kernel to current stream.

### CODE REVIEW

**Review by soumith — APPROVED** [review]

File: torch/csrc/jit/fusion_compiler.cpp [hunk 1]
```diff
@@ -550,13 +551,20 @@ struct CUDAFusionFunction : public CompiledFusionFunction {
...
      // cudaFree(0) accomplishes this.
-     cudaFree(0);
-
+     CUcontext pctx = 0;
+     TORCH_CU_CHECK(cuCtxGetCurrent(&pctx));
+     if (!pctx) {
+        std::unique_lock<std::mutex> cudaFreeMutexLock(
+        *(THCCachingAllocator_getCudaFreeMutex()));
+        cudaFree(0);
+        cudaFreeMutexLock.unlock();
```
Comment (apaszke) [inline torch/csrc/jit/fusion_compiler.cpp]: This is unnecessary. It’s going to be unlocked when it leaves the scope which happens just after this line

File: torch/csrc/jit/fusion_compiler.cpp [hunk 2]
```diff
@@ -550,13 +551,20 @@ struct CUDAFusionFunction : public CompiledFusionFunction {
      // it is possible that this is the first cuda call on this thread
      // so make sure we initialize the Driver API's context
      // cudaFree(0) accomplishes this.
-     cudaFree(0);
-
+     CUcontext pctx = 0;
+     TORCH_CU_CHECK(cuCtxGetCurrent(&pctx));
+     if (!pctx) {
+        std::unique_lock<std::mutex> cudaFreeMutexLock(
+        *(THCCachingAllocator_getCudaFreeMutex()));
```
Comment (apaszke) [inline torch/csrc/jit/fusion_compiler.cpp]: Can you please add an indent here? It looks as if it was a new independent line

**Comment by Oil3** [comment]
> NotImplementedError: The operator 'aten::upsample_bicubic2d.out' is not currently implemented for the MPS device. If **you want this op to be added in priority during the prototype phase of this feature, please comment on https://github.com/pytorch/pytorch/issues/77764. As a temporary fix, you can set the environment variable `PYTORCH_ENABLE_MPS_FALLBACK=1` to use the CPU as a fallback for this op. WARNING: this will be slower than running natively on MPS

What's actually needed to implement that for MPS?

**Comment by rachelcenter** [comment]
just got this error: 

Error occurred when executing AIO_Preprocessor:

The operator 'aten::upsample_bicubic2d.out' is not currently implemented for the MPS device. If you want this op to be added in priority during the prototype phase of this feature, please comment on https://github.com/pytorch/pytorch/issues/77764. As a temporary fix, you can set the environment variable `PYTORCH_ENABLE_MPS_FALLBACK=1` to use the CPU as a fallback for this op. WARNING: this will be slower than running natively on MPS.

**Pull request merged**